### PR TITLE
Dynamic input file generation for nvme and nvmf devices v2

### PR DESCRIPTION
**Pull request by manvanthar**
Code changes include
1. review comments from PR303
2. changes to differentiate FC and nvmf devices.
3. new logic to identify nvmf config and devices
4. logic to differentiate b/w nvmf RHEL8 and other Distros
5. correction of pci_devices values

### CODE REVIEW

**Comment by manvanthar** [comment]
#python3 ./pci_info.py --show-info
[{'adapter_description': 'Emulex Corporation LPe31000/LPe32000 Series '
                         '16Gb/32Gb Fibre Channel Adapter',
  'adapter_id': '10df:e300:1014:0615',
  'adapter_type': 'nvmf',
  'class': 'scsi_host',
  'disks': ['/dev/mapper/eui.a8000000000000e900507608108001b3',
            '/dev/mapper/eui.a8000000000000ea00507608108001b3',
            '/dev/mapper/eui.a8000000000000eb00507608108001b3',
            '/dev/mapper/eui.a8000000000000ec00507608108001b3',
            '/dev/mapper/eui.a8000000000000ed00507608108001b3'],
  'driver': ['nvme_fc', 'nvme_fabrics', 'nvme_core'],
  'firmware': '14.0.574.7',
  'functions': ['0014:70:00.0', '0014:70:00.1'],
  'infiniband_interfaces': [],
  'interfaces': ['host1', 'host2'],
  'is_root_disk': False,
  'mpath_disks': ['/dev/mapper/eui.a8000000000000e900507608108001b3',
                  '/dev/mapper/eui.a8000000000000ea00507608108001b3',
                  '/dev/mapper/eui.a8000000000000eb00507608108001b3',
                  '/dev/mapper/eui.a8000000000000ec00507608108001b3',
                  '/dev/mapper/eui.a8000000000000ed00507608108001b3'],
  'mpath_wwids': ['eui.a8000000000000e900507608108001b3',
                  'eui.a8000000000000ea00507608108001b3',
                  'eui.a8000000000000eb00507608108001b3',
                  'eui.a8000000000000ec00507608108001b3',
                  'eui.a8000000000000ed00507608108001b3'],
  'pci_root': '0014:70:00',
  'slot': 'U78D3.001.WZS067A-P1-C12'}]

#python3 ./pci_info.py --create-config
10:16:58 INFO    : python avocado-setup.py --run-suite host_io_nvmf_rhel8_0014_70_00_fvt --input-file io_input.txt
# cat io_input.txt 
[io_nvmf_rhel8_0014_70_00_fvt]
disk = "/dev/mapper/eui.a8000000000000e900507608108001b3"
lv_disks = "/dev/mapper/eui.a8000000000000ea00507608108001b3"
htx_disks = "/dev/mapper/eui.a8000000000000eb00507608108001b3"
disks = "/dev/mapper/eui.a8000000000000ec00507608108001b3 /dev/mapper/eui.a8000000000000ed00507608108001b3"
wwids = "eui.a8000000000000e900507608108001b3 eui.a8000000000000ea00507608108001b3 eui.a8000000000000eb00507608108001b3 eui.a8000000000000ec00507608108001b3 eui.a8000000000000ed00507608108001b3"
pci_device = "0014:70:00.0"
pci_devices = "0014:70:00.0 0014:70:00.1"
module = "nvme_fc nvme_fabrics nvme_core"
dir = "/mnt"
lv_size = "10G"
count = 30
fsstress_loop = 3
num_of_dlpar = 30
num_of_hotplug = 20
iteration = 30
run_type = "rpm"
manageSystem = ""
userid = ""
password = ""
hmc_username = ""
hmc_pwd = ""
htx_rpm_link = ""
switch_name = ""

File: lib/pci.py [hunk 1]
```diff
@@ -547,7 +690,7 @@ def pci_info(pci_addrs, pci_type='', pci_blocklist='', type_blocklist=''):
         pci_dic['is_root_disk'] = False
         for disk in pci_dic['disks']:
             for root_disk in root_disks:
-                if root_disk == disk:
+                if root_disk in disk:
```
[the remark this answers is not in the document]
Comment (abdhaleegit) [reply lib/pci.py:693]: any need to change it back ?

File: lib/pci.py [hunk 2]
```diff
@@ -188,8 +294,8 @@ def get_interfaces_in_pci_address(pci_address, pci_class):
     if not pci_class or not os.path.isdir(pci_class_path):
         return ""
     return [interface for interface in os.listdir(pci_class_path)
-            if os.path.islink(os.path.join(pci_class_path, interface)) and pci_address in os.readlink(os.path.join(pci_class_path, interface))]
-
+            if pci_address in os.readlink(os.path.join(pci_class_path,
+                                                       interface))]
```
[the remark this answers is not in the document]
Comment (abdhaleegit) [reply lib/pci.py:298]: any reason to revert back to old code ?

**Review by abdhaleegit — CHANGES_REQUESTED** [review]
rhel8x  has  /dev/mapper/mpathx has nvmf device format
rhel9x and sles15x  has /dev/nvme0n1 as nvmf device format

hope this is taken care for all above, currently we have io_nvmf_rhel8x.cfg and io_nvmf.cfg  --> for all os other than 8x

**Comment by abdhaleegit** [comment]
@manvanthar remove the revert commit and rebase your branch to master

File: lib/pci.py [hunk 3]
```diff
@@ -459,6 +571,25 @@ def get_primary_ioa(pci_address):
...
+    Gets the wwids of the nvmf multipath disks
+
+    :return: wwids
+    """
+    lines = runcmd('multipath -ll')[1].splitlines()
+    wwid_list = []
+    nvme_pattern = r"NVME"
+
+    # Regular expression pattern to match WWID
+    wwid_pattern = r"eui\.[a-fA-F\d]+"
```
Comment (PraveenPenguin) [inline lib/pci.py:585]: is `eui` always be wildcard?

File: lib/helper.py [hunk 4]
```diff
@@ -179,6 +180,21 @@ def copy_dir_file(src, dest):
...
+
+    :return: True if it is rhel8 False otherwise
+    """
+    system_info = platform.system()
+    release_info = platform.release()
+
+    if system_info == "Linux" and "el8" in release_info:
+        return True
+    else:
+        return False
```
[the remark this answers is not in the document]
Comment (abdhaleegit) [reply lib/helper.py:195]: I think we are good here, as the python platform module is actully reading /etc/os-release  and these values will not change with custom kernel

**Review by abdhaleegit — APPROVED** [review]
LGTM,  we can merge after praveens changes are pushed

File: lib/pci.py [hunk 5]
```diff
@@ -39,6 +39,120 @@ def get_domains():
...
+
+    :return: nvmf disks.
+    """
+    fc_host_names = []
+    node_names = []
+    nvme_names = []
+    nvmf_disks = []
+
+    sys_devices_path = '/sys/devices'
+    pci_domain_path = os.path.join(sys_devices_path, 'pci' + pci_domain)
```
[the remark this answers is not in the document]
Comment (manvanthar) [reply lib/pci.py]: Done

**Review by PraveenPenguin — APPROVED** [review]

**Pull request merged**